examples/pytorch/gcmc/train_sampling.py

```diff
@@ -271,47 +223,47 @@ def config():
 
     return args
 
-@thread_wrapped_func
+#@thread_wrapped_func
 def run(proc_id, n_gpus, args, devices, dataset):
     dev_id = devices[proc_id]
     train_labels = dataset.train_labels
     train_truths = dataset.train_truths
     num_edges = train_truths.shape[0]
-    sampler = GCMCSampler(dataset,
-                          'train')
 
-    seeds = th.arange(num_edges)
-    dataloader = DataLoader(
-        dataset=seeds,
+    reverse_types = {str(k): 'rev-' + str(k) for k in dataset.possible_rating_values}
+    reverse_types.update({v: k for k, v in reverse_types.items()})
+    sampler = dgl.dataloading.MultiLayerNeighborSampler([None])
+    dataloader = dgl.dataloading.EdgeDataLoader(
+        dataset.train_enc_graph,
+        {str(k): th.arange(dataset.train_enc_graph.number_of_edges(etype=str(k)))
+         for k in dataset.possible_rating_values},
+        sampler,
+        exclude='reverse_types',
+        reverse_etypes=reverse_types,
+        return_eids=True,
         batch_size=args.minibatch_size,
-        collate_fn=sampler.sample_blocks,
         shuffle=True,
-        pin_memory=True,
-        drop_last=False,
-        num_workers=args.num_workers_per_gpu)
+        drop_last=False)
 
     if proc_id == 0:
-        valid_sampler = GCMCSampler(dataset,
-                                    'valid')
-        valid_seeds = th.arange(dataset.valid_truths.shape[0])
-        valid_dataloader = DataLoader(dataset=valid_seeds,
-                                      batch_size=args.minibatch_size,
-                                      collate_fn=valid_sampler.sample_blocks,
-                                      shuffle=False,
-                                      pin_memory=True,
-                                      drop_last=False,
-                                      num_workers=args.num_workers_per_gpu)
-
-        test_sampler = GCMCSampler(dataset,
-                                   'test')
-        test_seeds = th.arange(dataset.test_truths.shape[0])
-        test_dataloader = DataLoader(dataset=test_seeds,
-                                     batch_size=args.minibatch_size,
-                                     collate_fn=test_sampler.sample_blocks,
-                                     shuffle=False,
-                                     pin_memory=True,
-                                     drop_last=False,
-                                     num_workers=args.num_workers_per_gpu)
+        valid_dataloader = dgl.dataloading.EdgeDataLoader(
+            dataset.valid_dec_graph,
+            th.arange(dataset.valid_dec_graph.number_of_edges()),
+            sampler,
+            g_sampling=dataset.valid_enc_graph,
```

Comment:
@jermainewang @classicsong I found GCMC a particularly interesting use case because in validation I would like to iterate over a validation edge set (`dataset.valid_dec_graph`), but perform neighborhood sampling on the graph formed by only the training edge set (`dataset.valid_enc_graph`).  So here I'm separating the graph argument into a graph argument whose edges are iterated over, and another graph argument whose neighborhoods are sampled.

Do you think it's OK?  I feel it's kinda tedious.

Reply:
I think this make sense.
We sample edges first and then neighbor sample nodes.